File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int result = 0;
        int minElementToEnd = INT_MAX;

        for (int end = 0; end < prices.size(); ++end) {
            minElementToEnd = min(prices[end], minElementToEnd);
            result = max(prices[end] - minElementToEnd, result);
        } 
        return result;
    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # 7 1 2 6 so the max profit is 6 - 1 = 5
       # sliding window?
        max_profit = 0
        l = 0
        
        for r in range(len(prices)):
            if prices[r] > prices[l]:
                max_profit = max(max_profit, prices[r] - prices[l])
            
            else:
                l = r
        
        return max_profit